sm/plugins/tedge_apt_plugin/src/main.rs

```diff
@@ -73,27 +101,33 @@ fn run(operation: PluginOp) -> Result<ExitStatus, InternalError> {
             status
         }
 
-        PluginOp::Install { module, version } => {
-            if let Some(version) = version {
-                run_cmd(
-                    "apt-get",
-                    &format!("install --quiet --yes {}={}", module, version),
-                )?
-            } else {
-                run_cmd("apt-get", &format!("install --quiet --yes {}", module))?
+        PluginOp::Install { module, version } => install(module, version)?,
+
+        PluginOp::Remove { module, version } => remove(module, version)?,
+
+        PluginOp::UpdateList => {
+            let mut updates: Vec<SoftwareModuleUpdate> = Vec::new();
+            let mut status: ExitStatus = ExitStatus::from_raw(0);
+            let mut rdr = csv::ReaderBuilder::new()
+                .has_headers(false)
+                .delimiter(b' ')
+                .flexible(true)
+                .from_reader(io::stdin());
+            for result in rdr.deserialize() {
+                updates.push(result?);
             }
-        }
 
-        PluginOp::Remove { module, version } => {
-            if let Some(version) = version {
-                // check the version mentioned present or not
-                run_cmd(
-                    "apt-get",
-                    &format!("remove --quiet --yes {}={}", module, version),
-                )?
-            } else {
-                run_cmd("apt-get", &format!("remove --quiet --yes {}", module))?
+            print!("{:?}", updates);
+            for update_module in updates {
+                status = match update_module.action {
+                    UpdateAction::Install => install(update_module.name, update_module.version)?,
+                    UpdateAction::Remove => remove(update_module.name, update_module.version)?,
+                };
+                if !status.success() {
+                    break;
+                }
             }
```

Comment:
I take this as a very first step.

* The list of packages to install and remove *must* be provided to apt in a *single* command invocation.
* Otherwise there is no point to support this command.
* An example is `sudo apt-get install --quiet --yes mosquitto-clients libmosquitto1-`  to install `mosquitto-clients` *and* to also remove `libmosquitto1`. This example *must* fail while installing `mosquitto-clients` then removing `libmosquitto1` will be successful with nothing installed.

Reply:
Passing every argument in a single command is allowed only when the packages are provided without an explicit version. If the packages are passed to `apt` with the `<package>=<version>` syntax, it doesn't support the trick with `-` at the end to indicate uninstall. That's why I had to resort to this method with `apt`.

Reply:
One can have something intermediate, though.

A version can be provided for the packages to be installed only - letting the version unset on remove.

For instance the following is working as expected.
```
$ sudo apt-get install --quiet --yes mosquitto-clients=1.5.7-1+deb10u1 libmosquitto1-
...
The following packages have unmet dependencies:
 mosquitto-clients : Depends: libmosquitto1 (= 1.5.7-1+deb10u1) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

So what I propose, it to:
* Install/remove all the packages in one command - specifying the version only for the installed packages.
* If a version is specified for a package to be removed.
    * either fallback to the previous behavior (simpler but might introduce surprises)
    * check the version installed and raised an error if different
* If a file is provided for a package to be installed
   *  fallback to the previous behavior
*

Reply:
It is doable but I doubt if it'll make any difference for this `apt` plugin. When the software removal requests come from the cloud, they will always carry a version as this `apt` plugin reports the installed packages to the cloud with versions. Since uninstall requests are derived from this reported list, they will always have the versions that the device reported. So, for the apt plugin, we'll always end up falling back to the non-updateList behaviour.

Reply:
I don't follow you. Since all the uninstall requests come with the version actually installed, we will *never* fall back to the non-updateList behavior.

* So, I think that for a first version we can simply ignore the versions provided along the modules to be removed.
* Furthermore, I do think that if the `apt` plugin doesn't support the update-list feature, then the whole feature is useless and this PR can be closed.

Reply:
> * If a version is specified for a package to be removed.
>    * either fallback to the previous behavior (simpler but might introduce surprises)

I was talking about this proposal to fallback if the module to be removed has a version. But, if we decide to ignore the versions in the removal requests (since those versions are anyway the ones reported by the device) it should be fine. We'll just do a local validation if the provided version in the incoming request is the same as the one installed to avoid accidentally removing some package that the user updated locally on the device bypassing the software management feature from the cloud.